File: oct15/G.cpp
#include <iostream>

using namespace std;

int main() {
	long long dict[55][55];
	long long N;

	for (int i = 0; i < 55; i++) {
		for (int j = 0; j <= i; j++) {
			if (j == 0)
				dict[i][j] = 1;
			else if (j == i)
				dict[i][j] = 1;
			else
				dict[i][j] = dict[i - 1][j - 1] + dict[i - 1][j];
		}
	}

	while (cin >> N) {
		int cnt[55];
		int i, j;
		long long ans = 0;

		if (N < 7) {
			cout << "Day " << N << ": Level = " << 0 << endl;
			continue;
		}

		for (i = 55; i >= 0; i--)
			if ((N & ((1 << i)) != 0) {
				cout << "N: " << N << ", " << i << ", " << (N & (1 << i)) << endl;
				break;
			}

		for (j = i - 1; j >= 0; j--) {
			cnt[j] = cnt[j + 1] + (N & (1 << j) != 0); 
		}

		for (j = 0; j <= i; j++) {
			if (N & (1 << j) != 0) {
				int k = (cnt[0] - cnt[j] + 2) % 3;
				while (k <= j) {
					ans += dict[j][k];
					k += 3;
				}
			}
		}

		cout << "Day " << N << ": Level = " << ans << endl;
	}
}

File: oct15/G.py
import sys


def countC(c, n):
	res = 1
	for i in xrange(c):
		res *= n - i
	for i in xrange(1, c+1):
		res /= i
	return res

def count3(x, n):
	res = 0
	while x <= n:
		res += countC(x, n)
		x += 3

	return res 

for line in sys.stdin:
	n = int(line.strip())

	b = "{0:b}".format(n)
	L = len(b)
	pos = [ L - i for i in xrange(L) if b[i] == "1" ]
	#print pos

	height = 0

	for i in xrange(len(pos)):
		c3 = (3 - i % 3) % 3
		temp = count3( c3 , pos[i]-1)
		height += temp
		#print "(", c3, pos[i]-1, ")", ":", temp

	if len(pos) % 3 == 0:
		height += 1

	height -= 1

	#print height
	print "Day {0}: Level = {1}".format(n, height)

	


